### DISCUSSION TITLE
Converting mass

### DISCUSSION THREAD

**Discussion by pakastin** [cosinekitty/astronomy#205] [Q&A] [answered]
Hi and thank you for an awesome library! I can't get my head around how to convert mass from kg to the unit that's in use in this library (trying to calculate gravity between a spaceship and planets).

**Comment by cosinekitty** [comment, marked as answer]
If you are trying to calculate the trajectory of a spacecraft in the Solar System, chances are you can treat the spacecraft's mass as negligible (it has no measurable effect on the planet). Then all you need to know is the acceleration experienced by the spacecraft due to the Sun and planets. Internally, I do this same thing to calculate Pluto's orbit: I treat Pluto as an insignificant mass that is accelerated by the Sun and the major planets (Jupiter, Saturn, Uranus, and Neptune).

The mass units I use (as returned by the `MassProduct` functions) are the product of the universal gravitation constant G with the mass of the Sun or planet M. The product GM is known much more accurately than G or M are separately. The units of GM are AU^3/day^2 (cubic astronomical units per squared day). These are nice units for expressing the inverse square law: a = GM/r^2 is the acceleration experienced by a small body that is r astronomical units away from the center of the mass M, where "a" is measured in AU/day^2.  You can convert to other units as needed.

If you really need the masses of solar system bodies in kg, you can multiply the return value of the `MassProduct` function by 6.7196e+33.

Does this help, or are you looking for something else?

**Reply by pakastin** [reply]
I just realized mass was not the only problem calculating accelerations, but also time, since i guess the time unit in astronomy.js is day 🤔

**Reply by cosinekitty** [reply]
Yes, that is correct! At some point in the future, I am pondering adding a generic gravitation integrator to Astronomy Engine, that could be used for calculating the trajectory of an asteroid, comet, or spacecraft passing through the solar system, given that small body's initial position vector and velocity vector. I would rework the Pluto calculator to use the generic function, so the code as a whole would not grow very much.

I learned the hard way from my Pluto experience that there is a lot more to this than is obvious at first. For one thing, everything has to be done in solar system barycentric coordinates, not heliocentric coordinates, because the Sun is also wobbling around (the major 4 planets pull on it more than you would think). The SSB (solar system barycenter) is an inertial frame of reference, in which Newtonian mechanics is easy to express. The center of the Sun is *not* a good frame of reference, LOL.

Another thing is that it is surprisingly difficult to minimize creeping errors in the mutual dependence between position and acceleration: the gravitational accelerations are changing continuously as everything moves, but the *way* everything moves depends on how the accelerations are changing. After a lot of research and experimentation, I finally found something that worked for Pluto with good accuracy. I could probably save people a lot of headache by providing this for arbitrary small bodies.

**Reply by pakastin** [reply]
I'm starting to understand why people say "it's not rocket science", when they're definitely not talking about rocket science 😅

**Comment by pakastin** [comment]
Thank you for helping me out! 🙏 Got physics and conversions working now and released alpha version of my free 2D space flight simulator: https://flanets.io

zoom out and you'll see all planets in the right place and spaceship gravity physics playing nicely with astronomy.js 👌

**Reply by cosinekitty** [reply]
That is very cool!

**Reply by cosinekitty** [reply]
Oh, by the way, I added #207 as a future feature goal. I don't know when I get to it, but it will help simulations like this.

**Reply by pakastin** [reply]
I used physics from here: http://www.braeunig.us/space/orbmech.htm#launch

**Reply by cosinekitty** [reply]
Very nice. It looks like fun bedtime reading... LOL!  Seriously, I will study that because it looks interesting.

**Reply by pakastin** [reply]
It is very interesting, and got quite a bit to get working. Then i realized i need to use atan2 and not atan function to get quadrants right 😂

**Reply by pakastin** [reply]
Well, and few other things 🤣

**Reply by pakastin** [reply]
Even more interesting stuff here: http://www.braeunig.us 😉

**Reply by cosinekitty** [reply]
Yes, atan2 is one of my favorite functions. Astronomy Engine provides some handy functions for converting between spherical and cartesian coordinates, by the way. In case they are useful for you: [VectorFromSphere](https://github.com/cosinekitty/astronomy/tree/master/source/js#VectorFromSphere) and [SphereFromVector](https://github.com/cosinekitty/astronomy/tree/master/source/js#SphereFromVector). There are all kinds of coordinate transform functions too. Let me know if there's anything I can help with.
